Sources/Signal.swift

```diff
@@ -391,6 +302,137 @@ public final class Signal<Value, Error: Swift.Error> {
+		///
+		/// # Thread Safety Notes
+		///
+		/// - Check if the signal is at a specific state.
+		///   * Read directly.
+		///
+		/// - Deliver `value` events with the alive state.
+		///   * `sendLock` must be acquired.
+		///
+		/// - Replace the alive state with another. (e.g. observers bag manipulation)
+		///   * `updateLock` must be acquired.
+		///
+		/// - Transition from `alive` to `terminating` as a result of receiving
+		///   a termination event.
+		///   * `updateLock` must be acquired, and should fail gracefully if the
+		///     signal has terminated.
+		///
+		/// - Check if the signal is terminating. If it is, invoke `tryTerminate`
+		///   which transitions the state from `terminating` to `terminated`, and
+		///   delivers the termination event.
+		///   * Both `sendLock` and `updateLock` must be acquired. The check can be
+		///     relaxed, but the state must be checked again after the locks are
+		///     acquired. Fail gracefully if the state has changed since the relaxed
+		///     read, i.e. a concurrent sender has already handled the termination
+		///     event.
+		///
+		/// Exploiting the relaxation of reads, please note that false positives
+		/// are intentionally allowed in the `terminating` checks below. As a
+		/// result, normal event deliveries need not acquire `updateLock`.
+		/// Nevertheless, this should not cause the termination event being
+		/// sent multiple times, since `tryTerminate` would not respond to false
+		/// positives.
+		private final var state: State
+
+		/// Used to ensure that reference counting of state updates do not race against
+		/// relaxed state reads.
+		///
+		/// The negative one is appropriated to indicate the `Signal` is being terminated.
+		private final let updaterCount: UnsafeAtomicInt32
+
+		/// Used to ensure that state updates are serialized.
+		final let updateLock: Lock
+
+		init() {
+			state = .alive(AliveState(observers: Bag()))
+			updaterCount = UnsafeAtomicInt32(0)
+			updateLock = Lock.make()
+		}
+
+		final var hasStartedTermination: Bool {
+			return updaterCount.swap(from: -1, to: -1)
+		}
+
+		/// Read the alive state with the given action, if the `Signal` is alive.
+		final func withAliveStateRelaxed(_ body: (AliveState) -> Void) {
+			// If we know that there are pending concurrent state update, `updateLock`
+			// needs to be acquired to avoid a concurrent release-retain race.
+			let updateLock: Lock? = updaterCount.value > 0 ? self.updateLock : nil
+
+			updateLock?.lock()
+
+			if case let .alive(state) = self.state {
+				updateLock?.unlock()
+				body(state)
+			} else {
+				updateLock?.unlock()
+			}
+		}
+
+		/// Initiate a state update.
+		final func updateState<R>(_ body: (inout State) -> R) -> R {
+			var count: Int32
+
+			repeat {
+				count = updaterCount.value
+				guard count >= 0 else {
+					assert(count == -1)
+					break
+				}
+			} while !updaterCount.swap(from: updaterCount.value, to: updaterCount.value + 1)
```

Comment:
This seems like a poor imitation of a spin/unfair lock. I'd prefer to see us use `Atomic` like in #453—at least to reestablish correctness. Afterwards we could consider an optimization like this.